pkg/dockerregistry/server/prune.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/docker/distribution/registry/storage/driver"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func Prune(ctx context.Context, storageDriver driver.StorageDriver, registry distribution.Namespace, registryClient RegistryClient) {
+	logger := context.GetLogger(ctx)
+
+	repositoryEnumerator, ok := registry.(distribution.RepositoryEnumerator)
+	if !ok {
+		logger.Fatal("unable to convert Namespace to RepositoryEnumerator")
+	}
+
+	oc, _, err := registryClient.Clients()
+	if err != nil {
+		logger.Fatalf("error getting clients: %s", err)
+	}
+
+	imageList, err := oc.Images().List(metav1.ListOptions{})
+	if err != nil {
+		logger.Fatalf("error listing images: %s", err)
+	}
+
+	inuse := make(map[string]string)
+	for _, image := range imageList.Items {
+		// Keep the manifest.
+		inuse[image.Name] = image.DockerImageReference
+
+		// Keep the config for a schema 2 manifest.
+		if image.DockerImageManifestMediaType == schema2.MediaTypeManifest {
+			inuse[image.DockerImageMetadata.ID] = image.DockerImageReference
+		}
+
+		// Keep image layers.
+		for _, layer := range image.DockerImageLayers {
+			inuse[layer.Name] = image.DockerImageReference
+		}
+	}
+
+	err = repositoryEnumerator.Enumerate(ctx, func(repoName string) error {
+		logger.Debugln("Processing repository", repoName)
+
+		named, err := reference.WithName(repoName)
+		if err != nil {
+			return fmt.Errorf("failed to parse repo name %s: %v", repoName, err)
+		}
+
+		repository, err := registry.Repository(ctx, named)
+		if err != nil {
+			return err
+		}
+
+		manifestService, err := repository.Manifests(ctx)
+		if err != nil {
+			return err
+		}
+
+		manifestEnumerator, ok := manifestService.(distribution.ManifestEnumerator)
+		if !ok {
+			return fmt.Errorf("unable to convert ManifestService into ManifestEnumerator")
+		}
+
+		err = manifestEnumerator.Enumerate(ctx, func(dgst digest.Digest) error {
+			if imageReference, ok := inuse[string(dgst)]; ok {
+				logger.Debugf("Keep manifest %s@%s (manifest belongs to image %s)", repoName, dgst, imageReference)
+				return nil
+			}
+
+			logger.Printf("Deleting manifest: %s@%s", repoName, dgst)
+			err := manifestService.Delete(ctx, dgst)
+			if err != nil {
+				return fmt.Errorf("delete manifest %s: %s", dgst, err)
+			}
+
+			return nil
+		})
+		if e, ok := err.(driver.PathNotFoundError); ok {
+			logger.Warnf("prune manifests in repository %s: %s", repoName, e)
+		} else if err != nil {
+			return fmt.Errorf("prune manifests in repository %s: %s", repoName, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	logger.Debugln("Processing blobs")
+	vacuum := storage.NewVacuum(ctx, storageDriver)
+	err = registry.Blobs().Enumerate(ctx, func(dgst digest.Digest) error {
+		if imageReference, ok := inuse[string(dgst)]; ok {
+			logger.Debugf("Keep blob %s (blob belongs to image %s)", dgst, imageReference)
+			return nil
+		}
+
+		err := vacuum.RemoveBlob(string(dgst))
```

Comment:
It is already logged from inside of RemoveBlob.
```
INFO[0000] Deleting blob: /docker/registry/v2/blobs/sha256/f6/f60ecfcf302b2378acfc896904b637497acb7863ded1c8c867c92e97593ae412  go.version=go1.8.3 instance.id=68dc9918-abae-47cc-b118-8f0f161c5cca
```
Should I log it one more time, but with the digest only?